### PR TITLE
Fixes #48 Add networking to wallet

### LINKED ISSUE #48: Backend to manage the networking
**Feature description:**
Need a backend to manage the networking. No image available (for now, check the electron app).

**Possible implementation:**

### PR DESCRIPTION
**Pull request by stdevMac**
Fixes #48 

- [x] Add networking interfaces
- [x] Add networking Manager for Qt binding
- [x] Backend for Networking

### CODE REVIEW

**Comment by olemis** [comment]
@stdevMac I tried to merge this , but I am not sure the result is correct . Could you please take a look ?

File: src/core/network.go [hunk 1]
```diff
@@ -4,4 +4,23 @@ type PEX interface {
 	getTxnPool() TransactionIterator
```
Comment (olemis) [inline src/core/network.go]: names of all methods of this interface must start with capital letters

File: src/core/network.go [hunk 2]
```diff
@@ -4,4 +4,23 @@ type PEX interface {
...
+	Value() INetwork
+	Next() bool
+	HasNext() bool
+}
+
+type NetworkSet interface {
+	ListNetworks() NetworkIterator
 }
+
+type INetwork interface {
```
Comment (olemis) [inline src/core/network.go]: `INetwork` => `PexNode`

File: src/core/network.go [hunk 3]
```diff
@@ -4,4 +4,23 @@ type PEX interface {
 	getTxnPool() TransactionIterator
 	getConnections() 
 	broadcast(txn Transaction)
+
+type NetworkIterator interface {
```
Comment (olemis) [inline src/core/network.go]: `NetworkIterator` => `PexNodeIterator`

File: src/core/network.go (hunk 1 above)
Comment (olemis) [reply src/core/network.go]: @stdevMac does it make sense to move this interface to core ?

File: pkg/networking/networking.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+import (
+	//"encoding/json"
+	//"fmt"
+	//"io/ioutil"
+	//"net/http"
+	"github.com/skycoin/skycoin/src/api"
+)
+
+//func GetDefaultConnections(NodeAddress string)  {
```
Comment (olemis) [inline pkg/networking/networking.go]: shall we keep it or remove it ?

File: pkg/networking/networking.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+func nodeUsername() string{
+	return ""
+}
+
+//----------------TODO
+func nodePassord() string{
+	return ""
+}
+
+func newClient() *api.Client{
```
Comment (olemis) [inline pkg/networking/networking.go]: this must be defined just once in `src/coin/skycoin/client.go`

Comment (olemis) [reply pkg/networking/networking.go]: or a similar file

File: pkg/networking/networkingModel.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
+package networking
+
+import (
+	"github.com/therecipe/qt/core"
+	//"github.com/skycoin/skycoin/src/api"
+	"github.com/skycoin/skycoin/src/readable"
+)
+
+const(
+	Peer	=	int(core.Qt__UserRole) + 1
```
Comment (olemis) [inline pkg/networking/networkingModel.go]: use `iota`

File: pkg/networking/networkingModel.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	Source	=	int(core.Qt__UserRole) + 2
+	BlockHeight	=	int(core.Qt__UserRole) + 3
```
Comment (olemis) [inline pkg/networking/networkingModel.go]: not aligned correctly ?

File: src/coin/skycoin/models/networking.go [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+import (
+	"log"
+	"strings"
+
+	"github.com/fibercrypto/FiberCryptoWallet/src/core"
+	"github.com/skycoin/skycoin/src/api"
+	"github.com/skycoin/skycoin/src/readable"
+)
+
+type SkycoinNetworkIterator struct {
```
Comment (olemis) [inline src/coin/skycoin/models/networking.go]: `SkycoinNetworkIterator` => `SkycoinPexNodeIterator`

File: src/coin/skycoin/models/networking.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+func NewSkycoinRemoteNetwork(nodeAddress string) *SkycoinRemoteNetwork {
+	return &SkycoinRemoteNetwork{nodeAddress}
+}
+
+func (remoteNetwork *SkycoinRemoteNetwork) newClient() *api.Client {
+	return api.NewClient(remoteNetwork.nodeAddress)
+}
+
+func (remoteNetwork *SkycoinRemoteNetwork) ListNetworks() core.NetworkIterator {
```
Comment (olemis) [inline src/coin/skycoin/models/networking.go]: Is this for P2P or API nodes ? P2P peers may not have enabled REST API .

Comment (olemis) [reply src/coin/skycoin/models/networking.go]: in any case `SkycoinRemoteNetwork` is a name I can't understand , should be changed

File: src/coin/skycoin/models/networking.go [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+	}
+	var netIterator []core.INetwork
+	for _, con := range nets.Connections {
+		netIterator = append(netIterator, connectionsToNetwork(con))
+	}
+
+	return NewSkycoinNetworkIterator(netIterator)
+}
+
+type Network struct {
```
Comment (olemis) [inline src/coin/skycoin/models/networking.go]: `Network` => `SkycoinPexNode`

File: src/coin/skycoin/models/networking.go [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+func (network Network) GetIp() string {
+	return network.Ip
+}
+
+func (network Network) GetPort() uint16 {
+	return network.Port
+}
+
+func (network Network) GetBlock() uint64 {
```
Comment (olemis) [inline src/coin/skycoin/models/networking.go]: `GetBlock` => `GetBlockHeight`

File: src/models/networkingModel.go [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+import (
+	"time"
+
+	coin "github.com/fibercrypto/FiberCryptoWallet/src/core"
+	"github.com/therecipe/qt/core"
+)
+
+// ip, port, source, block, lastSeenIn, lastSeenOut
+const (
+	Ip          = int(core.Qt__UserRole) + 1
```
Comment (olemis) [inline src/models/networkingModel.go]: use `iota`

File: src/ui/Networking.qml [hunk 13]
```diff
@@ -83,12 +83,36 @@ Page {
 
     } // Frame
 
+
     // Roles: ip, port, source, block, lastSeenIn, lastSeenOut
     // Implement the model in the backend (a more recommendable approach)
-    ListModel { // EXAMPLE
+    //ListModel { // EXAMPLE
```
Comment (olemis) [inline src/ui/Networking.qml]: does this need to be removed ?

**Review by olemis — CHANGES_REQUESTED** [review]

File: src/ui/Networking.qml (hunk 13 above)
Comment (stdevMac) [reply src/ui/Networking.qml]: yes, already removed, this was the QML initial examples

File: src/models/networkingModel.go (hunk 12 above)
Comment (stdevMac) [reply src/models/networkingModel.go]: done

File: src/coin/skycoin/models/networking.go (hunk 11 above)
Comment (stdevMac) [reply src/coin/skycoin/models/networking.go]: done

File: src/coin/skycoin/models/networking.go (hunk 10 above)
Comment (stdevMac) [reply src/coin/skycoin/models/networking.go]: done

File: src/coin/skycoin/models/networking.go (hunk 9 above)
Comment (stdevMac) [reply src/coin/skycoin/models/networking.go]: This connect to API nodes, About the name I don't now complete sure about the change I made, I rename it to `SkycoinNetworkConnections`, What do you think?

File: pkg/networking/networking.go (hunk 5 above)
Comment (stdevMac) [reply pkg/networking/networking.go]: files removed

File: pkg/networking/networking.go (hunk 4 above)
Comment (stdevMac) [reply pkg/networking/networking.go]: files removed

File: src/core/network.go (hunk 3 above)
Comment (stdevMac) [reply src/core/network.go]: done

File: src/core/network.go (hunk 2 above)
Comment (stdevMac) [reply src/core/network.go]: Done

File: src/core/network.go (hunk 1 above)
Comment (stdevMac) [reply src/core/network.go]: This interface isn't in use, Is just a path to follow once we have to do that. In this moment, we should see about that

File: pkg/networking/networkingModel.go (hunk 6 above)
Comment (stdevMac) [reply pkg/networking/networkingModel.go]: files removed

File: pkg/networking/networkingModel.go (hunk 7 above)
Comment (stdevMac) [reply pkg/networking/networkingModel.go]: files removed

File: src/coin/skycoin/models/networking.go (hunk 8 above)
Comment (stdevMac) [reply src/coin/skycoin/models/networking.go]: done

File: src/core/network.go (hunk 1 above)
Comment (stdevMac) [reply src/core/network.go]: What do you think about this @olemis

**Comment by olemis** [comment]
@stdevMac conflicts

**Review by olemis — APPROVED** [review]

**Pull request merged**